third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -518,15 +528,63 @@ impl LifetimeState {
         self.parent_edges(label).any(|(_, e)| e.kind.is_mut())
     }
 
-    /// Returns the children grouped by their edge kind.
-    fn grouped_children(
+    /// Returns the children of the given nodes, grouped into hyper edges. A hyper edge
+    /// is constituted by a set of edge kinds and associated list of edges. Each hyper edge
+    /// represents an abstract borrow operation.
+    ///
+    /// 1) All edges which lead into the same node are considered to be part of the same hyper
+    /// edge. Consider:
+    ///
+    /// ```text
+    ///           \     /
+    ///         e1 \   / e2
+    ///             \ /
+    ///              n
+    /// ```
+    /// This forms a hyper edge `{e1.kind, e2.kind} -> [e1, e2]`. Both edges have to be in the same
+    /// group because `n` has a 'weak' borrow history, it can either stem from `e1` or `e2`.
+    ///
+    /// 2) For all other edges not leading into the same node, they are grouped according
+    /// their kind. Consider:
+    ///
+    /// ```text
+    ///            |    |
+    ///         e1 |    | e2
+    ///            |    |
+    ///           n1   n2
+    /// ```
+    /// If `e1.kind == e2.kind`, this forms a hyper edge `{e1.kind} -> [e1, e2]`, otherwise it will
+    /// be two independent hyper edges `{e1.kind} -> [e1]` and `{e2.kind} -> [e2]`. The former
+    /// reflects that the edges of the same kind are the same abstract borrow operation, independent
+    /// of the number of edges involved.
+    fn group_children_into_hyper_edges(
         &self,
         labels: &BTreeSet<LifetimeLabel>,
-    ) -> BTreeMap<BorrowEdgeKind, Vec<&BorrowEdge>> {
-        let mut result: BTreeMap<BorrowEdgeKind, Vec<&BorrowEdge>> = BTreeMap::new();
-        for label in labels {
-            for edge in self.children(label) {
-                result.entry(edge.kind.clone()).or_default().push(edge)
+    ) -> BTreeMap<BTreeSet<BorrowEdgeKind>, Vec<&BorrowEdge>> {
+        // First ompute map from target nodes to edges, allowing to identify weak edges.
+        let mut target_to_incoming: BTreeMap<LifetimeLabel, Vec<&BorrowEdge>> = BTreeMap::new();
+        for edge in labels.iter().flat_map(|l| self.children(l)) {
+            target_to_incoming
+                .entry(edge.target)
+                .or_default()
+                .push(edge)
+        }
+        // Now compute the result.
+        let mut result: BTreeMap<BTreeSet<BorrowEdgeKind>, Vec<&BorrowEdge>> = BTreeMap::new();
+        for (_, mut edges) in target_to_incoming {
+            if edges.len() > 1 {
```

Comment:
If there are two labels in target_to_incoming that can generate the same key, is it possible to lose some edges in the result map because of the insert operation below?

Reply:
No, since we have `BTreeMap<LifetimeLabel, Vec<&BorrowEdge>>`, i.e. we collect all edges wiuth the same target in a vector.

Reply:
To be on the save side, I made this now such that edges with the same kind set are appended. I don't think it can happen but anyways.